### PR TITLE
Data Science jnyman

### PR DESCRIPTION
**Pull request by jnyman**
Similar to my attempt with Dino Catalog, I have once again used Hirb for display purposes.

My main focus was on the tests which all execute but I also included a runner file so that the "production data" could be executed. This also gave another way to run some of the test data sets.

### CODE REVIEW

**Comment by katieschilling** [comment]
Awesome job at making a clean git history. This PR is easy to read.

File: data_science/data/test_data_favors_b.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+[{"date":"2014-03-20","cohort":"B","result":1},
```
Comment (paulghaddad) [inline data_science/data/test_data_favors_b.json]: You probably don't need this many data points for your tests.

File: data_science/lib/test_data.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
+require 'abanalyzer'
+
+class TestData
+  attr_reader :data_sample, :group_name
+
+  # The value of 1.96 refers to the 95th percentile of a standard
+  # normal distribution. This is multiplied by the standard error.
+  CONFIDENCE_INTERVAL = 1.96
+
+  PROBABILITY_OF_CHANCE = 0.05
```
Comment (paulghaddad) [inline data_science/lib/test_data.rb]: Good use of constants here.

File: data_science/lib/test_data.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  # observed number of that outcome. Then square the result and divide that
+  # squared value by the expected number. Then you sum those values across all
+  # possible outcomes and you get the chi-square statistic.
+  def goodness_of_fit
+    groups = {}
+
+    group_variants.each do |variant|
+      groups[variant] = {
+        conversions: conversions_count(variant),
+        non_conversions: trial_size(variant) - conversions_count(variant) }
```
Comment (paulghaddad) [inline data_science/lib/test_data.rb]: The ending brace should be on a new line for styling purposes.

File: data_science/split_test_analyzer.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env ruby
```
Comment (paulghaddad) [inline data_science/split_test_analyzer.rb]: I like this!

File: data_science/split_test_analyzer.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    extend Hirb::Console
+
+    data_source = DataLoader.new(data)
+    @data = TestData.new(data_source.load_data, 'cohort')
+    data_table
+  end
+
+  def data_table
+    puts
+    table(display_data,
```
Comment (paulghaddad) [inline data_science/split_test_analyzer.rb]: This is difficult to read. Is there a better way to do this?

File: data_science/split_test_analyzer.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    percents = Hash[data.variant_percents]
+    percentage = (percents[variant] * 100).round(3)
+    standard_error = (data.standard_error(variant) * 100).round(3)
+
+    "Variant #{variant}: #{percentage}% +/- #{standard_error}%"
+  end
+
+  def display_favored_data
+    puts "\nFinal Analysis:"
+    puts @data.favored_variant
```
Comment (paulghaddad) [inline data_science/split_test_analyzer.rb]: I'd prefer the `puts` to be used when you call the method. For example, line 51 can read:

`puts analyzer.display_favored_data`

File: data_science/spec/test_data_favors_none_spec.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
+require 'data_loader'
```
Comment (paulghaddad) [inline data_science/spec/test_data_favors_none_spec.rb]: I'd prefer for your spec file to mirror the name of your Ruby files, such as `test_data_spec.rb`.

File: data_science/spec/test_data_favors_none_spec.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+require 'test_data'
+
+RSpec.describe TestData do
+  context 'Test Data Favors No Variant' do
```
Comment (paulghaddad) [inline data_science/spec/test_data_favors_none_spec.rb]: Nice use of a context, but you can combine this file and any other spec files that test `TestData`. Then use contexts to distinguish the various cases, such as the data not favoring any spec.

File: data_science/spec/test_data_favors_none_spec.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+      expect(test_data.trial_size('A')).to eq(20)
+    end
+
+    it 'will determine the number of conversions in a group' do
+      expect(test_data.conversions_count('B')).to eq(6)
+      expect(test_data.conversions_count('A')).to eq(11)
+    end
+
+    it 'will determine all group variants' do
+      expect(test_data.group_variants).to eq(%w(A B))
```
Comment (paulghaddad) [inline data_science/spec/test_data_favors_none_spec.rb]: Consider using this matcher: 

`expect(actual).to match_array(expected_array)`

See: https://github.com/rspec/rspec-expectations

File: data_science/spec/test_data_favors_none_spec.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    it 'will calculate the conversion percentage of group members' do
+      expect(test_data.conversion_percentage('B')).to eq(0.4)
+      expect(test_data.conversion_percentage('A')).to eq(0.55)
+    end
+
+    it 'will calculate the standard error' do
+      expect(test_data.standard_error('B')).to eq(0.2479225685572009)
```
Comment (paulghaddad) [inline data_science/spec/test_data_favors_none_spec.rb]: Consider using this matcher:

`expect(actual).to be_within(delta).of(expected)`

File: data_science/spec/test_data_favors_none_spec.rb (hunk 7 above)
Comment (jnyman) [inline data_science/spec/test_data_favors_none_spec.rb]: Yup, this is one of the big "divides" between developers and testers. Even for unit test files, testing discipline tends to suggest calling test files by declarative names related to their intent. That being said, I do know that in the Ruby community there is a huge debate about that. This is something where I would clearly do whatever seems to be the practice here.

File: data_science/spec/test_data_too_small_spec.rb [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      expect(test_data.conversion_percentage('A')).to eq(0.5)
+    end
+
+    it 'will calculate the standard error' do
+      expect(test_data.standard_error('B')).to eq(0.6929646455628166)
+      expect(test_data.standard_error('A')).to eq(0.6929646455628166)
+    end
+
+    it 'will calculate the goodness-of-fit' do
+      expect { test_data.goodness_of_fit }.to raise_error(
```
Comment (paulghaddad) [inline data_science/spec/test_data_too_small_spec.rb]: Good job testing your error condition!

**Comment by paulghaddad** [comment]
Overall, this is excellent work. Take a look at some of the specialized matchers I suggested. You are good to move on!

File: data_science/spec/test_data_favors_none_spec.rb (hunk 7 above)
Comment (paulghaddad) [inline data_science/spec/test_data_favors_none_spec.rb]: Good point! I do feel the practice here as Enova is mirroring the file name.

**Pull request closed without merging**